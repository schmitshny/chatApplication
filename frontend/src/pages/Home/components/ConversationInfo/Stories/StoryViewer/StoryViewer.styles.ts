import styled, { keyframes } from 'styled-components';

export const StoryContainer = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
`;

export const ImageContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  position: relative;
  height: 80vh;
  border-radius: 10px;
  overflow: hidden;
`;

export const StoryImage = styled.img`
  width: 500px;
  height: 100%;
  object-fit: cover;
`;

export const NavigationButton = styled.button`
  top: 50%;
  transform: translateY(-50%);
  border: none;
  cursor: pointer;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  background: none;
  &:disabled {
    visibility: hidden;
  }
`;

export const PreviousButton = styled(NavigationButton)`
  margin-right: 10px;
`;

export const NextButton = styled(NavigationButton)`
  margin-left: 10px;
  transform: translateY(-50%) rotate(180deg);
`;

export const UserInfo = styled.div`
  display: flex;
  align-items: center;
  color: white;
  position: absolute;
  top: 10px;
  left: 10px;
  padding: 4px 8px;
`;

export const UserName = styled.span`
  font-size: 16px;
  color: white;
  margin-left: 10px;
  text-shadow: 2px 2px 4px rgba(0, 0, 0, 0.5);
`;

const shrink = keyframes`
  from {
    width: 100%;
  }
  to {
    width: 0%;
  }
`;

export const ProgressBar = styled.div<{ $duration: number }>`
  height: 5px;
  background-color: white;
  animation: ${shrink} ${({ $duration }) => $duration / 1000}s linear forwards;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
`;
