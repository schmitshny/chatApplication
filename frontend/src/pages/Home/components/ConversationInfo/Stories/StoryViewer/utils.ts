export const resetAndStartTimer = (
  timerRef: React.MutableRefObject<number | undefined>,
  updateStoryIndex: () => void,
  duration: number,
  cleanupFunction: (timer: number | undefined) => void,
) => {
  cleanupFunction(timerRef.current);
  timerRef.current = window.setTimeout(() => {
    updateStoryIndex();
  }, duration);
};
