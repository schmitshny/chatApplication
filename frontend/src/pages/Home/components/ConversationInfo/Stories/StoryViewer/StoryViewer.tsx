import { useState, useEffect, useRef } from 'react';
import { Avatar, Icon } from '../../../../../../components';
import { Story } from '../../../../../../features/stories/types';
import {
  NextButton,
  PreviousButton,
  StoryContainer,
  StoryImage,
  UserInfo,
  UserName,
  ImageContainer,
  ProgressBar,
} from './StoryViewer.styles';
import { resetAndStartTimer } from './utils';

interface StoryViewerProps {
  stories: Story[];
  initialStoryIndex: number;
  onClose?: () => void;
  setActiveStoryId?: (id: number) => void;
}

export const STORY_DURATION_MS = 5000;

export const StoryViewer = ({ stories, initialStoryIndex, onClose, setActiveStoryId }: StoryViewerProps) => {
  const [currentStoryIndex, setCurrentStoryIndex] = useState(initialStoryIndex);
  const timerRef = useRef<number>();

  useEffect(() => {
    resetAndStartTimer(
      timerRef,
      () => {
        if (currentStoryIndex < stories.length - 1) {
          setCurrentStoryIndex(currentStoryIndex + 1);
          if (setActiveStoryId) setActiveStoryId(currentStoryIndex + 1);
        } else {
          onClose && onClose();
        }
      },
      STORY_DURATION_MS,
      clearTimeout,
    );

    const currentTimer = timerRef.current;
    console.log('initialStoryIndex', initialStoryIndex);

    return () => {
      if (currentTimer) {
        clearTimeout(currentTimer);
      }
    };
  }, [currentStoryIndex, stories.length, onClose, setActiveStoryId, initialStoryIndex]);

  useEffect(() => {
    setCurrentStoryIndex(initialStoryIndex);
  }, [initialStoryIndex]);

  const handleNextStory = () => {
    if (currentStoryIndex < stories.length - 1) {
      setCurrentStoryIndex(currentStoryIndex + 1);
      if (setActiveStoryId) setActiveStoryId(currentStoryIndex + 1);
    }
  };

  const handlePreviousStory = () => {
    if (currentStoryIndex > 0) {
      setCurrentStoryIndex(currentStoryIndex - 1);
      if (setActiveStoryId) setActiveStoryId(currentStoryIndex - 1);
    }
  };

  const currentStory = stories[currentStoryIndex];

  if (!currentStory) {
    // Możesz tutaj obsłużyć brak historii, np. wyświetlając komunikat błędu lub zamykając widok
    console.error('No story found at this index:', currentStoryIndex);
    return null; // lub inne zachowanie, jakie chcesz zaimplementować
  }

  const user = currentStory.user;

  return (
    <StoryContainer data-testid="story-viewer">
      <PreviousButton onClick={handlePreviousStory} disabled={currentStoryIndex === 0} data-testid="next-button">
        <Icon name="prev" customColor="#7393b3" />
      </PreviousButton>
      <ImageContainer>
        <ProgressBar $duration={STORY_DURATION_MS} key={currentStoryIndex} />
        <UserInfo>
          <Avatar avatarUrl={user.avatarImg} />
          <UserName>{`${user.name} ${user.lastName}`}</UserName>
        </UserInfo>
        <StoryImage src={currentStory.imageUrl} alt="Story" />
      </ImageContainer>
      <NextButton
        onClick={handleNextStory}
        disabled={currentStoryIndex === stories.length - 1}
        data-testid="previous-button"
      >
        <Icon name="prev" customColor="#7393b3" />
      </NextButton>
    </StoryContainer>
  );
};
