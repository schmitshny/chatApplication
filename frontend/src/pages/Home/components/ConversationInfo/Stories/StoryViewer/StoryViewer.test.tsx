import { describe, test, expect, vi } from 'vitest';
import { act, fireEvent, screen } from '@testing-library/react';
import { render } from '../../../../../../tests/test-utils';
import { STORY_DURATION_MS, StoryViewer } from './StoryViewer';
import { mockStories } from '../../../../../../features/stories/tests/mocks';

describe('StoryViewer', () => {
  test('renders the story viewer with initial story', () => {
    const { container } = render(<StoryViewer stories={mockStories} initialStoryIndex={0} onClose={vi.fn()} />);

    const expectedUserName = `${mockStories[0].user.name} ${mockStories[0].user.lastName}`;

    expect(screen.getByTestId('story-viewer')).toBeInTheDocument();
    expect(screen.getByText(expectedUserName)).toBeInTheDocument();
    expect(screen.getByAltText('Story')).toHaveAttribute('src', mockStories[0].imageUrl);
    expect(container).toMatchSnapshot();
  });

  test('navigates to next and previous story', () => {
    render(<StoryViewer stories={mockStories} initialStoryIndex={0} onClose={vi.fn()} />);

    let expectedUserName = `${mockStories[0].user.name} ${mockStories[0].user.lastName}`;

    fireEvent.click(screen.getByTestId('next-button'));
    expect(screen.getByText(expectedUserName)).toBeInTheDocument();

    expectedUserName = `${mockStories[1].user.name} ${mockStories[1].user.lastName}`;

    fireEvent.click(screen.getByTestId('previous-button'));
    expect(screen.getByText(expectedUserName)).toBeInTheDocument();
  });

  test('auto advances to the next story', async () => {
    vi.useFakeTimers({ shouldAdvanceTime: true });
    render(<StoryViewer stories={mockStories} initialStoryIndex={0} onClose={vi.fn()} />);

    const expectedNextUserName = `${mockStories[1].user.name} ${mockStories[1].user.lastName}`;

    act(() => {
      vi.advanceTimersByTime(STORY_DURATION_MS);
    });

    expect(screen.getByText(expectedNextUserName)).toBeInTheDocument();
  });
});
