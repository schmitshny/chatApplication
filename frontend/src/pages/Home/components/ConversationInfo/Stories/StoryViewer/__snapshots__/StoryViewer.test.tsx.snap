// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StoryViewer > renders the story viewer with initial story 1`] = `
<div>
  <div
    class="sc-ksZauW iVoqlh"
  >
    <div
      class="sc-hAZoWV igRyVY"
    >
      <div
        class="sc-bZkfhe fmQblX"
        data-testid="story-viewer"
      >
        <button
          class="sc-jIYZa-D sc-himrge fpJqVK eKnvTy"
          data-testid="next-button"
          disabled=""
        >
          <div
            class="sc-jqUWZQ bnkMjB"
          >
            <svg
              fill="none"
              height="34"
              id="Layer_1"
              version="1.1"
              viewBox="0 0 512 512"
              width="34"
              xml:space="preserve"
              xmlns="http://www.w3.org/2000/svg"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            >
              <g>
                <g>
                  <path
                    d="M256,0C114.837,0,0,114.837,0,256s114.837,256,256,256s256-114.837,256-256S397.163,0,256,0z M313.749,347.584
			c8.341,8.341,8.341,21.824,0,30.165c-4.16,4.16-9.621,6.251-15.083,6.251c-5.461,0-10.923-2.091-15.083-6.251L176.917,271.083
			c-8.341-8.341-8.341-21.824,0-30.165l106.667-106.667c8.341-8.341,21.824-8.341,30.165,0s8.341,21.824,0,30.165L222.165,256
			L313.749,347.584z"
                  />
                </g>
              </g>
            </svg>
          </div>
        </button>
        <div
          class="sc-kLLYlN heVQgd"
        >
          <div
            class="sc-jdANtv eqVMle"
          />
          <div
            class="sc-cCsQgl keulDg"
          >
            <div
              class="sc-jSMdHm eJItaO"
            >
              <img
                alt="user avatar"
                class="sc-gKXNOb gUEVCF"
                loading="lazy"
                shape="circle"
                src="http://example.com/avatar.jpg"
              />
            </div>
            <span
              class="sc-ciZfDS eOsdSL"
            >
              John Doe
            </span>
          </div>
          <img
            alt="Story"
            class="sc-ikZqro fkMjdu"
            src="http://example.com/story1.jpg"
          />
        </div>
        <button
          class="sc-jIYZa-D sc-gXmRPE fpJqVK OKXYv"
          data-testid="previous-button"
        >
          <div
            class="sc-jqUWZQ bnkMjB"
          >
            <svg
              fill="none"
              height="34"
              id="Layer_1"
              version="1.1"
              viewBox="0 0 512 512"
              width="34"
              xml:space="preserve"
              xmlns="http://www.w3.org/2000/svg"
              xmlns:xlink="http://www.w3.org/1999/xlink"
            >
              <g>
                <g>
                  <path
                    d="M256,0C114.837,0,0,114.837,0,256s114.837,256,256,256s256-114.837,256-256S397.163,0,256,0z M313.749,347.584
			c8.341,8.341,8.341,21.824,0,30.165c-4.16,4.16-9.621,6.251-15.083,6.251c-5.461,0-10.923-2.091-15.083-6.251L176.917,271.083
			c-8.341-8.341-8.341-21.824,0-30.165l106.667-106.667c8.341-8.341,21.824-8.341,30.165,0s8.341,21.824,0,30.165L222.165,256
			L313.749,347.584z"
                  />
                </g>
              </g>
            </svg>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
